### ISSUE TITLE
Panache: Simplify setting the tenant identifier

### ISSUE THREAD

**Issue by gbourant** [quarkusio/quarkus#41805]
### Description

Since the [recent merge](https://github.com/quarkusio/quarkus/pull/41518), Quarkus Panache now uses Hibernate Session instead of EntityManager.

It would be beneficial if Quarkus Panache offered a simpler method to set the tenant identifier, such as `Todos.withTenantId('admin').findAll()`.

**Comment by gastaldi** [comment]
The problem is that these finder methods are static, so it may be hard to achieve this syntax. It could work if we introduce some bytecode manipulation here, but I am not 100% sure.

**Comment by FroMage** [comment]
Oh. How is that related to the change from `EntityManager` to `Session`?

How do you currently select Hibernate tenants?

I can keep this in mind for Panache 2 that I'm currently working on.

**Comment by gbourant** [comment]
> Oh. How is that related to the change from `EntityManager` to `Session`?

Idk, I just thought that it would make my suggestion easier since the implementation would not have to worry about the EntityManager anymore.

> How do you currently select Hibernate tenants?

I use the following way
[https://stackoverflow.com/questions/76985783/how-to-query-from-multiple-tenants-of-hibernate6-in-a-single-request/76990686#76990686](https://stackoverflow.com/questions/76985783/how-to-query-from-multiple-tenants-of-hibernate6-in-a-single-request/76990686#76990686)

**Comment by FroMage** [comment]
OK thanks, that helps.

So, in order to use tenants, you have the following steps:

- A Quarkus REST filter that extracts a tenant id from query parameters
- This sets a `ThreadLocal`
- Then you open a new transaction, in order to make ORM create a new `Session` via the `SessionFactory`, associated with the new transaction
- That creation then (immediately or lazily) creates a new `Session` by querying your `TenantResolver` which uses the `ThreadLocal`
- Finally your new transaction has the session with the proper tenant.

Now, it appears from the code samples that you can create your transaction, and only later set the `ThreadLocal` for the tenant, which implies that creating a transaction does not create a session (otherwise it would be too late to set the tenant id), and that tells me that the entire API is fishy.

The tenant ID is bound to the `Session`, and I think this is bound to the lifetime of the current transaction. I also suspect we can't have two sessions for the same transaction. Certainly not if they're bound to the transaction. Perhaps one after the other, via manual operations, but certainly not two in parallel.

Just to illustrate, consider the following code, adapted from your sample's answer:

```java
QuarkusTransaction.requiringNew().run(() -> {
        TenantUtil.setTenantId("123");
        System.out.println(roleRepository.select().fetch()); // lazy session creation with tenant 123
        TenantUtil.setTenantId("456");
        System.out.println(roleRepository.select().fetch()); // I strongly suspect this will keep using tenant 123
    });
```

So… it appears to me that the tenant should be bound to the transaction and that the API should instead be something like:

```java
QuarkusTransaction.requiringNew().forTenant("123").run(() -> {
        System.out.println(roleRepository.select().fetch());
    });
```

This would not lure people into thinking they can change tenant after te session is created, because it's lazy and tricky and a side-effect of operation ordering.

So, although I could have given you an easier way to set the tenant in Panache 2, such as :

```java
QuarkusTransaction.requiringNew().run(() -> {
        System.out.println(Role.forTenant("123").listAll());
    });
```

This would still have the same confusing issue of what's supposed to happen if you do:

```java
QuarkusTransaction.requiringNew().run(() -> {
        System.out.println(Role.forTenant("123").listAll());
        System.out.println(Role.forTenant("456").listAll());
    });
```

Where our options for the second call are:

- Throw an exception
- Open a new session just for that call, but then we have two open sessions in parallel for the current transaction, and worse: nobody's hanging on to the second one so I suspect things will break in extremely confusing ways.
- Silently ignore it (the current problematic behaviour)

Therefore, I think that if tenants are tied to transactions, they should be set on transactions. For Hibernate ORM/Panache we don't have transaction APIs, so that's not where it fits. For Hibernate Reactive/Panache, we do have transaction/session management APIs, so that's somewhere we could make tenant IDs explicit.

WDYT @yrodiere @Sanne ?

**Comment by yrodiere** [comment]
Binding the tenant to the repository/entity class is both limited (what about multi-entity transactions?) and dangerous (see your examples) IMO. I wouldn't allow that.

Binding the tenant to the transaction makes sense IMO.

API-wise... your suggestion  (`QuarkusTransaction.requiringNew().forTenant("123")`) makes sense from the user's point of view, but implementation would get technical, as tenants only make sense for Hibernate ORM right now (though we've talked about extending them to datasources in #11861). Wouldn't it be misleading people into thinking the whole transaction is using tenant-specific resources, while in fact only Hibernate ORM operations are, and everything else (filesystem, direct JDBC access) is free to escape tenancy limits?

I suppose a more conservative approach, pending a more integrated handling of multitenancy in Quarkus, would be some helper related to Hibernate ORM/Reactive only, like so:

```
io.quarkus.hibernate.orm.runtime.tenant.TenantResolver.withTenant("foobar", () -> {
    // Some code block
})
```

This would:

* Check there is no active request/transaction (you must start the transaction *within* the code block)
* Override the tenant identifier for the duration of the code block execution.

Though to be honest, to me the whole thing screams "you should be using user impersonation instead", e.g. `Security.asUser("john@foobar").do(() -> { ... })` (I'm making this API up, but there's probably something like it).
After all, isn't the choice of a tenant a direct result of some sort of authentication, or at least identification?

**Comment by FroMage** [comment]
> Binding the tenant to the repository/entity class is both limited (what about multi-entity transactions?) and dangerous (see your examples) IMO. I wouldn't allow that.
> 
> Binding the tenant to the transaction makes sense IMO.

Right, so I didn't get this wrong. Good :)

> API-wise... your suggestion (`QuarkusTransaction.requiringNew().forTenant("123")`) makes sense from the user's point of view, but implementation would get technical, as tenants only make sense for Hibernate ORM right now (though we've talked about extending them to datasources in #11861). Wouldn't it be misleading people into thinking the whole transaction is using tenant-specific resources, while in fact only Hibernate ORM operations are, and everything else (filesystem, direct JDBC access) is free to escape tenancy limits?

I'm pretty sure FS access is not tied to the transaction, and I also suspect ORM (or HR) is the main user of a transaction. As for JDBC… OK sure.

> I suppose a more conservative approach, pending a more integrated handling of multitenancy in Quarkus, would be some helper related to Hibernate ORM/Reactive only, like so:
> 
> ```
> io.quarkus.hibernate.orm.runtime.tenant.TenantResolver.withTenant("foobar", () -> {
>     // Some code block
> })
> ```
> 
> This would:
> 
> * Check there is no active request/transaction (you must start the transaction _within_ the code block)
> * Override the tenant identifier for the duration of the code block execution.

Would it also start a transaction?

> Though to be honest, to me the whole thing screams "you should be using user impersonation instead", e.g. `Security.asUser("john@foobar").do(() -> { ... })` (I'm making this API up, but there's probably something like it). After all, isn't the choice of a tenant a direct result of some sort of authentication, or at least identification?

I did wonder about that. I'm pretty sure authentication is related, somehow. So, you have to make sure that your tenant ID depends on the logged in user. Directly (fetching the tenant ID from the user, which is kinda special if your user info comes from the DB) or indirectly (say, using a URL host name or prefix, whose access in any case has to be validated against the logged in user).

So, I don't think there's any "impersonating" going on in most cases, but you're right that it's tied to the current user.

**Comment by yrodiere** [comment]
> Would it also start a transaction?

No; from what you quoted: "(you must start the transaction within the code block)"

> So, I don't think there's any "impersonating" going on in most cases, but you're right that it's tied to the current user.

In every case where the tenant can be inferred from a HTTP request or current user, you can do so already, by implementing `io.quarkus.hibernate.orm.runtime.tenant.TenantResolver` -- see https://quarkus.io/guides/hibernate-orm#multitenancy.

That's why I'm saying "impersonating": if you need to set the tenant ID explicitly, it's necessarily because there is no HTTP request or current user to resolve the tenant ID. What you want is really to act "as if" a user of a given tenant was logged in; and in security, that's called impersonating.

**Comment by FroMage** [comment]
> > Would it also start a transaction?
> 
> No; from what you quoted: "(you must start the transaction within the code block)"

Ah, so that's two calls adding to nesting, not ideal.

> In every case where the tenant can be inferred from a HTTP request or current user, you can do so already, by implementing `io.quarkus.hibernate.orm.runtime.tenant.TenantResolver` -- see https://quarkus.io/guides/hibernate-orm#multitenancy.
> 
> If you need to set the tenant ID explicitly, it's necessarily because there is no HTTP request or current user to resolve the tenant ID from. That's why I'm saying "impersonating": what you want is really to act "as if" a user of a given tenant was logged in; and in security, that's called impersonating.

OK, I understand what you mean now, thanks.

So, that's probably rare to want to switch tenant within your method. But even then, intuitively, that tenant selection naturally "fits" at the point where we're creating the transaction.

**Comment by gbourant** [comment]
Also maybe add a way to execute a query without setting the tenant id.

For example:

```java
QuarkusTransaction.requiringNew().asRoot().run(() -> {
  System.out.println(roleRepository.select().fetch()); // returns all roles 
});
```

`io.quarkus.hibernate.orm.runtime.tenant.TenantResolver` already has support for that

```java
    /**
     * Does the given tenant id represent a "root" tenant with access to all partitions?
     *
     * @param tenantId a tenant id produced by {@link #resolveTenantId()}
     *
     * @return true is this is root tenant
     */
    default boolean isRoot(String tenantId) {
        return false;
    }
```
